src/fragment-generation-utils.js

```diff
@@ -1339,6 +1391,32 @@ const expandRangeStartToWordBound = (range) => {
   }
 };
 
+/**
+ * Moves the range edges to the first and last visible text nodes inside of it.
+ * @param {Range} range - the range to be modified
+ */
+const moveRangeEdgesToTextNodes = (range) => {
+  const firstNode = getFirstNodeForBlockSearch(range);
+  if (!isText(firstNode)) {
```

Comment:
Do we need this check? `getFirstNodeForBlockSearch` already early-returns if the first node is a visible text node.

Same on line 1410 below

Reply:
For what I understand getFirstNodeForBlockSearch can return both text and non text nodes.
 
 ```
const getFirstNodeForBlockSearch = (range) => {
  // Get a handle on the first node inside the range. For text nodes, this
  // is the start container; for element nodes, we use the offset to find
  // where it actually starts.
  let node = range.startContainer;
  if (node.nodeType == Node.ELEMENT_NODE &&
      range.startOffset < node.childNodes.length) {
    node = node.childNodes[range.startOffset];
  }
  return node;
};
```
That's why I'm checking to determine if we need to change the range start or not. Let me know if that makes sense.

Reply:
Right, that makes sense, but the first two lines of `getFirstTextNode` are identical (except the visibility check):
```
  const firstNode = getFirstNodeForBlockSearch(range);
  if (isText(firstNode) && fragments.internal.isNodeVisible(firstNode)) {
    return firstNode;
  }
```
so having a check here is duplicating work

Reply:
I didn't see that as an issue given that we're just comparing a property of the node to a constant. 
I can change it to something like:
```
if (firstNode !== firstTextNode) {
// update range start or early return if no text nodes
}

Reply:
Ahhh wait, now I see what I was missing: we want to avoid calling `range.setStart` and changing the offset to 0 if it already points to a text node.

I don't feel strongly about it anymore, either the existing code or what you described in your comment is fine by me.

Reply:
I actually like it better after the refactoring :) 
```
const moveRangeEdgesToTextNodes = (range) => {
  const firstTextNode = getFirstTextNode(range);
  // No text nodes in range. Collapsing the range and early return.
  if (firstTextNode == null) {
    range.collapse();
    return;
  }

  const firstNode = getFirstNodeForBlockSearch(range);

  // Making sure the range starts with visible text. 
  if (firstNode !== firstTextNode) {
    range.setStart(firstTextNode, 0);
  }

  const lastNode = getLastNodeForBlockSearch(range);
  const lastTextNode = getLastTextNode(range);
  // No need for no text node checks here because we know at theres at least
  // firstTextNode in the range.

  // Making sure the range ends with visible text.
  if (lastNode !== lastTextNode) {
    range.setEnd(lastTextNode, lastTextNode.textContent.length);
  }
};
```